File: Scripts/FuncWall.cs

// Externally set in inspector per instance
float speed = 0.64f; // save
Vector3 targetPosition;
FuncStates startState; // save
float percentAjar = 0; // save
float percentMoved = 0; // save
FuncStates currentState; // save
int[] chunkIDs; // save
float startTime; // save
Vector3 startPosition; // save
Rigidbody rbody;
bool stopSoundPlayed; // save

void InitializeFromLoad() {
    rbody = GetComponent<Rigidbody>();
    SFXSource = GetComponent<AudioSource>();
    rbody.collisionDetectionMode = CollisionDetectionMode.ContinuousSpeculative;
    Vector3 tempVec = Vector3_A_minus_B(instances[i].position, targetPosition);
    float distTotal = distance_vector3(startPosition, targetPosition);
    tempVec = -tempVec.normalized;
    if (currentState == FuncStates_AjarMovingTarget)     tempVec *= (distTotal * percentAjar);
    else if (currentState == FuncStates_AjarMovingStart) tempVec *= (distTotal * (1f - percentAjar));
    else if (currentState == FuncStates_MovingStart)     tempVec *= (distTotal * (1f - percentMoved));
    else                                                 tempVec *= (distTotal * percentMoved);

    tempVec.x = vclamp(tempVec.x,-10000.0f,10000.0f);
    tempVec.y = vclamp(tempVec.y,-10000.0f,10000.0f);
    tempVec.z = vclamp(tempVec.z,-10000.0f,10000.0f);
    tempVec += instances[i].position;
    instances[i].position = tempVec;
}

File: Scripts/funcwall.c
#include "mod.h"

void FuncWallInitAfterLoad(uint16_t self) {
    Entity* e = &Eng_Global->instances[self];
    Vector3 tempVec = Vector3_A_minus_B(e->position,e->targetPosition);
    float distTotal = distance_vector3(e->startPosition,e->targetPosition);
    tempVec = scale_vector3(normalize_vector3(tempVec),-1.0f);
    if (e->funcState == FuncStates_AjarMovingTarget) tempVec = scale_vector3(tempVec,distTotal * e->percentAjar);
    else if (e->funcState == FuncStates_AjarMovingStart) tempVec = scale_vector3(tempVec,distTotal * (1.0f - e->percentAjar));
    else if (e->funcState == FuncStates_MovingStart) tempVec = scale_vector3(tempVec,distTotal * (1.0f - e->percentMoved));
    else tempVec = scale_vector3(tempVec,distTotal * e->percentMoved);
    e->position = Vector3_A_plus_B(e->position,tempVec);
}

void FuncWallMoveStart(uint16_t self) { Eng_Global->instances[self].funcState = FuncStates_MovingStart; Eng_Global->instances[self].tickFinished = Eng_Global->pauseRelativeTime + 10.0f; }
void FuncWallMoveTarget(uint16_t self) { Eng_Global->instances[self].funcState = FuncStates_MovingTarget; Eng_Global->instances[self].tickFinished = Eng_Global->pauseRelativeTime + 10.0f; }

void FuncWallTargetted(uint16_t self, uint16_t activator, const char* argvalue) {
    (void)activator; (void)argvalue;
    Entity* e = &Eng_Global->instances[self];
    if (e->funcState == FuncStates_Start || e->funcState == FuncStates_MovingStart || e->funcState == FuncStates_AjarMovingTarget) FuncWallMoveTarget(self);
    else FuncWallMoveStart(self);
    play_wav(sounds[76],1.0f,e->position,true);
    flag_set(&e->entflags,ENTFLAG_STOPSOUND_PLAYED,false);
}

void FuncWallUpdate(uint16_t self) {
    Entity* e = &Eng_Global->instances[self];
    Vector3 goal = e->funcState == FuncStates_MovingStart ? e->startPosition : e->targetPosition;
    FuncStates doneState = e->funcState == FuncStates_MovingStart ? FuncStates_Start : FuncStates_Target;
    if (e->funcState == FuncStates_Start) { e->position = e->startPosition; e->velocity = (Vector3){0.0f,0.0f,0.0f}; e->percentMoved = 0.0f; return; }
    if (e->funcState == FuncStates_Target) { e->position = e->targetPosition; e->velocity = (Vector3){0.0f,0.0f,0.0f}; e->percentMoved = 1.0f; return; }
    if (e->funcState != FuncStates_MovingStart && e->funcState != FuncStates_MovingTarget) return;
    Vector3 delta = Vector3_A_minus_B(goal,e->position);
    float distanceLeft = magnitude_vector3(delta);
    float total = distance_vector3(e->startPosition,e->targetPosition);
    float dist = e->speed * (float)Eng_Global->deltaTime;
    if (distanceLeft <= dist || e->tickFinished < Eng_Global->pauseRelativeTime) {
        e->position = goal;
        e->funcState = doneState;
        e->percentMoved = doneState == FuncStates_Target ? 1.0f : 0.0f;
        e->velocity = (Vector3){0.0f,0.0f,0.0f};
        return;
    }
    if (distanceLeft > 0.0001f) e->position = Vector3_A_plus_B(e->position,scale_vector3(normalize_vector3(delta),dist));
    if (total > 0.0001f) e->percentMoved = distance_vector3(e->startPosition,e->position) / total;
}
